### ISSUE TITLE
Some e-mails are unreadable after decryption

### ISSUE THREAD

**Issue by webratte** [thunderbird/thunderbird-android#1252]
### Expected behaviour

Mail should be readable
### Actual behaviour

See screenshot
### Steps to reproduce

I don't know when this will happens.
Some mails shown correct, some mails like on screenshot.
![k9](https://cloud.githubusercontent.com/assets/6864851/14231880/3bc1cdec-f994-11e5-987e-d9139e3dd8aa.png)
### Environment

Environment

K-9 Mail version:
5.108 (current alpha)

Android version:
5.0

Account type (IMAP, POP3, WebDAV/Exchange):
IMAP

**Comment by philipwhiuk** [comment]
Without knowing the difference between the two in terms of headers etc, this may be impossible to fix. At a guess it's message encoding related.

**Comment by Valodim** [comment]
Does this happen with an email you send to yourself?

**Comment by webratte** [comment]
No, not yet.
I will do some tests.
Maybe I will find similarities between this mails.

**Comment by mruwek** [comment]
Same here. Problematic e-mails are written (in my case) with Thunderbird on Linux x86_64:

> User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:38.0) Gecko/20100101 Thunderbird/38.6.0

However, there are some e-mails from the same sender that are decoded correctly. In both cases text is properly decrypted in mutt and charset is set to utf-8. There is one noticeable difference: properly decoded e-mails are encoded with **quoted-printable** and those with garbage text (similar to that above) are encoded with **base64**.

**Comment by webratte** [comment]
I use a service of www.mailbox.org what unencrypted mails encrypt by reach the Server.
With most mails I have no problem.
The Mails with this issue are from one company (first encrypted by mailbox.org)

Is there a mail address where I can send a (particular makes anonymous) header of a Mail with this problem? 
I don't know if there still sensible informations into the header. So I don't like to make it public.

If you check (and clean it from sensible data, feel free to post it here.

I hope my (google-) english ist good enough to understand me ;-)

**Comment by craftyguy** [comment]
What additional information is required to further debug this issue? I seem to also be experiencing this..

**Comment by craftyguy** [comment]
Here are two headers, one from an email that is correctly displayed in K-9 after decrypting, and the other that is displayed incorrectly.

PLEASE let me know if I can provide any more information to help here.. this issue has made K-9 rather unreliable for displaying encrypted emails for me, since about 20% of the emails I receive fall into the "K-9 can't handle it" category and are totally unreadable...


[notworking_header.txt](https://github.com/k9mail/k-9/files/250922/notworking_header.txt)
[working_header.txt](https://github.com/k9mail/k-9/files/250924/working_header.txt)

**Comment by Valodim** [comment]
@craftyguy do you have mails where this happens where the content is not sensitive? if you do, can you send me one of those, and add the output of `gpg --decrypt` over the contained `---BEGIN PGP MESSAGE---` part? That way I can re-encrypt that part thus reconstructing the entire message (but encrypted to a key I have). That would be tremendously helpful.

btw it seems like this happens to mailbox.org users more than others? Might just be a coincidence...

**Comment by craftyguy** [comment]
@Valodim, sure! 
I'm not totally sure what you mean by this though, can you point me to some instructions on how to do this?
"add the output of gpg --decrypt over the contained ---BEGIN PGP MESSAGE--- part"

**Comment by Valodim** [comment]
Tracked it down. The problem is that transfer-encoding is applied twice, once during decryption and once afterwards when the text is extracted.

@cketti you very explicitly applied this decoding during decryption: https://github.com/k9mail/k-9/blob/master/k9mail/src/main/java/com/fsck/k9/mailstore/DecryptStreamParser.java#L79

From the description of `Body.getInputStream` I really can't tell whether the data is supposed to be encoded for the receiving end or not:
```
    /**
     * Returns the raw data of the body, without transfer encoding etc applied.
     * TODO perhaps it would be better to have an intermediate "simple part" class where this method could reside
     *    because it makes no sense for multiparts
     */
```

I would read it that the data should not have transfer-encoding applied to it, but `MimeUtility.decodeBody` obviously expects it to be. So uh... yeah. Simple fix (just tested, works for me) is not decoding the message stream in DecryptStreamParser, but I'll wait for info on this one.

**Comment by webratte** [comment]
Sounds great.
Will you upload a Alpha with this fix to Play Store? 
I would like to test it too.

Thanx

**Comment by Valodim** [comment]
So um, unfortunately the fix breaks viewing of attachments. This is probably why cketti decoded the data in the first place: the attachment file provider expects the data in the file to be unencoded. two ways to go from here:

1. Teach the file provider how to deal with encoded data. This is a +100 lines solution and performance 
2. Change DecryptedTempFileBody to implement Body instead of RawDataBody. At the moment, we don't use the raw data for anything, so this would be a -5 lines solution, albeit slightly less nice conceptually.

I assume we'll go with 2), fix see referenced PR.

**Comment by craftyguy** [comment]
I can confirm that the issue is totally resolved for me now in 5.109! Y'all rock!
